### PR TITLE
test(hydration): Ruta A — P0-A/B/C + P1-A/B (reliability sprint)

### LINKED ISSUE #23: test(hydration): P0-A — hydration verifiable test matrix
## Summary

Ruta A (Confiabilidad Primero) — Bloque P0-A: Semanas 1–2.

Implementa la matriz contractual de tests de hidratación que valida el comportamiento de `commitHydrate` contra todos los casos críticos identificados en el blueprint.

## Problem

El código de hidratación (`commitHydrate`, markers, event reattachment) está implementado en `src/commit.ts` y `src/app.ts`, pero sin cobertura contractual. Solo existe 1 test básico de ~12+ escenarios necesarios.

## Proposed Solution

Agregar suite dedicada `tests/hydration.test.ts` con matriz completa:

- **Markers**: SSR incluye `data-axiom-id` en orden de recorrido correcto.
- **Happy path**: `commitHydrate` reusa nodos sin mismatch y re-adjunta eventos.
- **Mismatch tag**: detecta y reporta `Tag mismatch` en modo soft.
- **Mismatch text**: detecta y reporta `Text mismatch` en modo soft.
- **Mismatch children count**: detecta `Children count mismatch` en modo soft.
- **Strict mode**: lanza excepción en el primer mismatch con `strictMismatch: true`.
- **Portal missing (soft)**: warning con `skipMissingPortals: true`.
- **Portal missing (strict)**: throw `Portal target missing` con `skipMissingPortals: false`.
- **Debug payload**: `debug: true` exporta `__AXIOM_HYDRATION_DEBUG__` globalmente.
- **Deep tree**: árbol de 60 nodos hidrata sin mismatches.

Extender `tests/integration.test.ts` con flujo SSR → hydrate → interacción reactiva.
Ajustar `tests/app.test.ts` para claridad en caso root vacío + hydrate.

## Affected Files

- `tests/hydration.test.ts` — NEW
- `tests/integration.test.ts` — extend
- `tests/app.test.ts` — minor adjustment

## Acceptance Criteria

- [ ] Suite nueva pasa completa (0 failed).
- [ ] No regresión en suite existente.
- [ ] Sin cambios de runtime (solo tests).

### PR DESCRIPTION
**Pull request by naml14**
## Summary

Implements Ruta A phases P0-A through P1-B per blueprint (docs/RUTA-A-CONFIABILIDAD-PRIMERO.md).

Closes #23

## Changes

**P0-A: Hydration Verification Test Matrix**
- tests/hydration.test.ts (NEW, 350+ LOC)
- tests/app.test.ts & tests/integration.test.ts (extended)
- tests/benchmark.test.ts (extended with per-phase thresholds)

**P0-B: Error Context & Diagnostics**
- src/app.ts: Error envelope with phase tracking ('hydrate' distinct from 'commit')
- AppErrorContext now includes nodeKey field

**P0-C: Performance Baselines**
- CI regression thresholds for 1000-node trees
- Small tree (<50ms) and hydration benchmarks

**P1-A: Production Documentation**
- docs/SSR-HYDRATION-CONTRACT.md (NEW, 300 LOC)
- docs/TROUBLESHOOTING.md (NEW, 250 LOC) 
- README.md: SSR & Hydration section, Limitations table

**P1-B: Public Testing Utilities**
- src/testing.ts (NEW, 280 LOC): render(), fireEvent(), query helpers
- docs/TESTING-GUIDE.md (NEW, 380 LOC): comprehensive guide + 12+ examples

## Test Results

- 233/233 tests passing ✅
- No regressions in existing suite ✅
- typecheck clean ✅
- All CI gates green ✅

## Target Release

0.2.8 (reliability hardening sprint)

## Summary by Sourcery

Refuerza la fiabilidad de la hidratación SSR con nuevas pruebas, contexto de errores, umbrales de rendimiento, documentación y ayudantes públicos de testing.

New Features:
- Expone utilidades públicas de testing para componentes, incluyendo helpers de renderizado, consultas al DOM y disparo de eventos.
- Documenta el contrato de SSR e hidratación, los flujos de resolución de problemas y los patrones de testing para consumidores externos.

Bug Fixes:
- Garantiza que los errores relacionados con la hidratación se informen con una fase dedicada de hidratación (`hydrate`) en lugar de `commit` cuando corresponda.

Enhancements:
- Añade pruebas extensas centradas en la hidratación que cubren generación de marcadores, modos de manejo de desajustes, portales, árboles profundos y flujos de integración.
- Amplía los benchmarks de rendimiento con umbrales por fase y específicos de hidratación tanto para árboles grandes como pequeños.
- Enriquece el contexto de errores con un campo opcional `nodeKey` para diagnósticos más granulares.
- Aclara el README con una sección de uso de SSR & Hydration y una tabla explícita de limitaciones.
- Mejora las utilidades de configuración del DOM de pruebas para un testing de integración SSR-a-hidratación más fiable.

Documentation:
- Añade una referencia detallada del contrato de SSR & hidratación, una guía de resolución de problemas y una guía de testing alineada con las nuevas utilidades.

Tests:
- Introduce una suite de pruebas dedicada a hidratación que cubre la reutilización SSR en el camino feliz, clases de desajuste, portales, payloads de depuración y árboles profundos.
- Aumenta las pruebas de integración para validar el comportamiento de extremo a extremo SSR → hydrate → interacción y los casos límite de hidratación con hot-reload.
- Endurece y amplía las pruebas de benchmark con umbrales aptos para CI para `prepare`, `reflow`, `commit`, árboles pequeños y escenarios de hidratación.

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Strengthen SSR hydration reliability with new tests, error context, performance thresholds, docs, and public testing helpers.

New Features:
- Expose public testing utilities for components, including render helpers, DOM queries, and event dispatch.
- Document the SSR and hydration contract, troubleshooting flows, and testing patterns for external consumers.

Bug Fixes:
- Ensure hydration-related errors are reported with a dedicated hydrate phase instead of commit where applicable.

Enhancements:
- Add extensive hydration-focused tests covering marker generation, mismatch handling modes, portals, deep trees, and integration flows.
- Extend performance benchmarks with per-phase and hydration-specific thresholds for both large and small trees.
- Enrich error context with an optional nodeKey field for finer-grained diagnostics.
- Clarify README with an SSR & Hydration usage section and an explicit limitations table.
- Improve test DOM setup utilities for more reliable SSR-to-hydration integration testing.

Documentation:
- Add a detailed SSR & hydration contract reference, a troubleshooting guide, and a testing guide aligned with the new utilities.

Tests:
- Introduce a dedicated hydration test suite covering happy-path SSR reuse, mismatch classes, portals, debug payloads, and deep trees.
- Augment integration tests to validate end-to-end SSR → hydrate → interaction behavior and hot-reload hydration edge cases.
- Tighten and expand benchmark tests with CI-friendly thresholds for prepare, reflow, commit, small trees, and hydration scenarios.

</details>

### CODE REVIEW

File: src/testing.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  document.body.appendChild(container)
+
+  // Mount the app
+  const app = createApp(component, container, {
+    lineHeight: options?.lineHeight ?? 20,
+    font: options?.font ?? '16px sans-serif',
+    textEngine: options?.textEngine,
+  })
+
+  let isMounted = false
```
[the remark this answers is not in the document]
Comment (naml14) [reply src/testing.ts]: Resuelto

File: src/app.ts [hunk 2]
```diff
@@ -410,15 +412,22 @@ export function createApp(
...
+            throw err
+          }
         } else {
           commitFull(layout, prepared, root, state.domState)
         }
       } catch (err) {
-        reportError(err, resolveContextFromPrepared('commit', cycle, prepared))
+        if (!((err instanceof Error) && (options?.hydrate === true))) {
+          reportError(err, resolveContextFromPrepared('commit', cycle, prepared))
+        }
```
[the remark this answers is not in the document]
Comment (naml14) [reply src/app.ts]: Resuelto

File: docs/TESTING-GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+
+```ts
+import { defineComponent } from 'axiom-framework'
+import { render } from 'axiom-framework/testing'
+
+const Greeting = defineComponent(() => ({
+  tag: 'div',
+  children: [
+    { tag: 'h1', text: 'Hello' },
+    { tag: 'p', text: 'Welcome to axiom' },
```
[the remark this answers is not in the document]
Comment (naml14) [reply docs/TESTING-GUIDE.md]: Resuelto

File: src/testing.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+  /** Low-level: access the underlying App instance (advanced use only) */
+  _app: App
+}
+
+/**
+ * Render a component into a temporary DOM environment suitable for unit testing.
+ *
+ * Automatically sets up happy-dom, creates a root container, mounts the component,
+ * and provides query helpers.
```
[the remark this answers is not in the document]
Comment (naml14) [reply src/testing.ts]: Resuelto

File: docs/TROUBLESHOOTING.md [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+renderToString(MyComponent, { rootId: 'main' })
+
+// client
+const root = document.getElementById('main')!
+createApp(root, MyComponent, { hydrate: true })
+```
+
+**Component returns `null` or throws during prepare**
+
+Fix: Wrap `app.mount()` in try/catch and check `err.context.phase` to isolate the failure.
```
[the remark this answers is not in the document]
Comment (naml14) [reply docs/TROUBLESHOOTING.md]: Resuelto

File: docs/SSR-HYDRATION-CONTRACT.md [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+const root = document.getElementById('app')!
+const app = createApp(root, MyComponent, { hydrate: true })
+await app.mount()
+```
+
+**Advanced / testing use via `commitHydrate` directly:**
+
+```ts
+import { prepare } from 'axiom-framework'
+import { reflow } from 'axiom-framework/reflow'
```
[the remark this answers is not in the document]
Comment (naml14) [reply docs/SSR-HYDRATION-CONTRACT.md]: Resuelto

File: docs/SSR-HYDRATION-CONTRACT.md [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+- Portals: `skipMissingPortals` still applies even in strict mode
+
+Use strict mode in production to catch stale HTML caches, CDN mismatch, or
+component drift between server and client builds.
+
+### Debug mode
+
+`hydrationDebug: true` / `debug: true`
+
+Emits a `console.debug` line for every node processed:
```
[the remark this answers is not in the document]
Comment (naml14) [reply docs/SSR-HYDRATION-CONTRACT.md]: Resuelto

File: src/testing.ts [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
+}
+
+/**
+ * Capitalize first letter.
+ * @internal
+ */
+function capitalize(str: string): string {
+  return str.charAt(0).toUpperCase() + str.slice(1)
+}
```
[the remark this answers is not in the document]
Comment (naml14) [reply src/testing.ts]: Resuelto

**Comment by naml14** [comment]
Auditoría de comentarios PR #24 completada ✅

- Revisé todos los hilos de revisión (38/38) y no hay conversaciones sin resolver.
- Verifiqué los hallazgos activos/no-outdated contra el código actual:
  - `src/testing.ts`: `render()` monta correctamente, `isMounted` se actualiza, `rerender()` implementado, `queryByText()` mejorado.
  - `src/app.ts`: evitada doble notificación `hydrate`/`commit` para errores no-`Error`.
  - `package.json`: export subpath `./testing` presente.
  - `docs/TESTING-GUIDE.md`: ejemplo de `data-testid="form-input"` consistente.
- Comprobaciones locales ejecutadas:
  - `bun run typecheck` ✅
  - `bun test` ✅ (296 pass, 2 skip, 0 fail)
- Status checks de GitHub para PR #24: en verde ✅

Conclusión: no quedan correcciones pendientes derivadas de comentarios de revisión en esta PR.
